frontend/js/maze1.js

```diff
@@ -0,0 +1,297 @@
+
+            this.startPoint.innerHTML = '0';
+            this.endPoint.dataset.info = '';
+
+            this._step = 1;
+        }
+
+        if (!this.endPoint.innerHTML) {
+            let wasMarked = this._waveExpansionStep();
+
+            if (!wasMarked) {
+                message = 'There is no path could be found!';
+            }
+        } else {
+            this._backTraceStart();
+
+            if (!this._backTraceStep()) {
+                message = 'The path was found!';
+            }
+        }
+
+        if (message) {
+            this._finish(message);
+        }
+    }
+
+    cellNeighbours(y,x) {
+        let neighbours = [];
+
+        for (let i = Math.max(y-1, 0); i <= Math.min(y+1, this.size-1); i++) {
+            for (let j = Math.max(x-1, 0); j <= Math.min(x+1, this.size-1); j++) {
+                if ( this._neighborhoodCondition(x,y,j,i) ) {
+                    neighbours.push(this._map.rows[i].cells[j]);
+                }
+            }
+        }
+
+        return neighbours;
+    }
+
+//------------- event handlers----------------
+    static _changeCell(event) {
+        let td = event.target.closest('td');
+
+        if (!td) {
+            return;
+        }
+
+        if (td.dataset.info) {
+            td.dataset.info = '';
+        } else {
+            td.dataset.info = 'W'
+        }
+
+        Maze._toggleWall(td);
+    }
+
+    _markPoint(event) {
+        let td = event.target.closest('td');
+
+        if (!td) {
+            return;
+        }
+
+        if (this._choosing.point) {
+            if (td === this._choosing.point) {
+                this._choosing.point = null;
+
+                td.dataset.info = '';
+                this._choosing.toggle(td);
+            }
+        } else if (!td.dataset.info) {
+            this._choosing.point = td;
+
+            td.dataset.info = this._choosing.info;
+            this._choosing.toggle(td);
+        }
+
+        switch (this._choosing.info) {
+            case 'S': {
+                this._startPoint = this._choosing.point;
+                break;
+            }
+            case 'E': {
+                this._endPoint = this._choosing.point;
+                break;
+            }
+        }
+    }
+
+//------------- private methods---------------
+    _renderMaze() {
+        let mazeHtml = '<table data-selector="map">';
+
+        for(let i = 0; i<this.size; i++) {
+            mazeHtml +='<tr>\n';
+            for (let j = 0; j<this.size; j++) {
+                mazeHtml+='<td></td>\n';
+            }
+            mazeHtml +='</tr>';
+        }
+
+        mazeHtml +=  '</table>';
+        this._el.insertAdjacentHTML('beforeEnd',mazeHtml);
+    }
+
+    _waveExpansionStep() {
+        let step = this._step;
+        let wasMarked = false;
+
+        for (let i = 0; i < this.size; i++) {
+            for (let j = 0; j < this.size; j++) {
+                let td = this._map.rows[i].cells[j];
```

Comment:
you use HTML logic of a table 
but what if you decided to change markup?
you should work with array instead
and render it when needed